pyccel/codegen/printing/ccode.py

```diff
@@ -346,7 +345,7 @@ def is_c_pointer(self, a):
             return True
         if isinstance(a, FunctionCall):
             a = a.funcdef.results[0].var
-        if isinstance(getattr(a, 'dtype', None), CustomDataType) and a.is_argument:
+        if isinstance(getattr(a, 'dtype', None), CustomDataType) and (a.is_argument or a.get_user_nodes(FunctionDef)):
```

Comment:
Please can you explain this change

Reply:
The `a.get_user_nodes(FunctionDef)` checks if the variable passed in the function `is_c_pointer` is in a `Method Def`. All the attributes of a class in a `Method Def` should be pointers.

Reply:
> All the attributes of a class in a `Method Def` should be pointers.

I disagree, but that is not what this does. It ensures that all instances of a class in a `Method Def` should be pointers. E.g:
```
class Point1:
    def __init__(self, x):
        self.x = x

class Point2:
    def __init__(self, y):
        self.y = y

    def test_func(self):
        p = Point1(self.y)
        print(p.x)
```
You say that `p` has to be a pointer?

Reply:
Blocked by https://github.com/pyccel/pyccel/pull/1479

Reply:
Fixed in https://github.com/pyccel/pyccel/pull/1472/commits/4f440b59370f1170d0f5e188b881ea0b41c9bcf8.
But the `self.y` in `test_func` isn't working cause  his `get_user_nodes(ClassDef)` returns `None` but I think normally it has to return the `ClassDef(Point2)`.

Reply:
It can't return `ClassDef(Point2)` until the function `__init__` has been added to the class (which only happens once it's built). I think you'll have to get the class def from the self variable or via the dtype

Reply:
Why do you need to know the current class?

Reply:
`self` is already an argument

Reply:
> 

So I can compare if it s the same class I ll return True in the `is_c_pointer`

Reply:
But it's already an argument. Doesn't it already return true?

Reply:
I m talking about `self.y` in `p = Point1(self.y)`

Reply:
Yes. `self.y` shouldn't be a pointer because `y` isn't a pointer, `self` should be a pointer because it is an argument

Reply:
So in c we will print `self->y` (a variable accessed via a pointer) not `&self->y` (a pointer accessed via a pointer)

Reply:
The `self` return `False` when I print the `is_argument`.

Reply:
I found a simple way and it s to add an attribute to the `CCodePrinter` and it s `self._current_class` and by using it I ll compare it with the cls_base of the variable passed in the `is_c_pointer`.

Reply:
> The `self` return `False` when I print the `is_argument` in all the `DottedVariable` in the exemple.

Ah. That's either means that there are multiple `self` variables or that there's a bug in the initialisation (`self` didn't used to be treated as an argument, but I changed it because it looks like one in most languages and has the same properties). Probably this should be fixed

Reply:
Can you take a look for what I changed. I think the `self._current_class` will really help me in the future.

Reply:
I have had a look. I imagine this works and I think `_current_class` will be useful, but I still think it could be useful to fix the `is_argument` property.

Or maybe we should just open an issue to remove it and use something like `expr.has_user(FunctionDefArgument)` instead

Reply:
Fixed https://github.com/pyccel/pyccel/pull/1472/commits/b9b353587bd848a608c5e2150cef6434decb65f0.